### PR TITLE
feat: skip block cache deallocation to make shutdown fast

### PR DESCRIPTION
**Pull request by wanghenshui**
when blockcache size is too big, skip dtor could save shutdown time

ref https://groups.google.com/g/rocksdb/c/xQ_o9jWoFqg

### CODE REVIEW

**Comment by torwig** [comment]
@wanghenshui You can run `./x.py format` before committing changes to format the code.

**Review by git-hulk — DISMISSED** [review]

File: src/storage/storage.cc [hunk 1]
```diff
@@ -98,6 +98,7 @@ void Storage::CloseDB() {
   db_->SyncWAL();
   rocksdb::CancelAllBackgroundWork(db_.get(), true);
   for (auto handle : cf_handles_) db_->DestroyColumnFamilyHandle(handle);
+  shared_block_cache_->DisownData();
```
Comment (mapleFU) [inline src/storage/storage.cc]: Should we add another flag for this? Since I don't know CloseDb in testing or other case would checking leak

Comment (wanghenshui) [reply src/storage/storage.cc]: ok then.

`fast_shutdown`?

Comment (mapleFU) [reply src/storage/storage.cc]: this is ok to me

File: kvrocks.conf [hunk 2]
```diff
@@ -643,6 +643,13 @@ migrate-batch-size-kb 16
 # Default: 16M
 migrate-batch-rate-limit-mb 16
 
+
+# Specifies fast shutdown mode
+# If set to 1, we skip blockcache dtor to speed up shutdown
+#
+# Default: 0
+# fast_shutdown 0
```
Comment (PragmaTwice) [inline kvrocks.conf]:
```suggestion
# skip-block-cache-deallocation-on-close no
```

**Review by PragmaTwice — COMMENTED** [review]
Also we need to add a go test case for it.

**Review by mapleFU — COMMENTED** [review]
`PrepareRestoreDB()` also call `CloseDB`, should it prevent from unmanage memory?

**Comment by wanghenshui** [comment]
> `PrepareRestoreDB()` also call `CloseDB`, should it prevent from unmanage memory?

yes. i 'll make a single function from `CloseDB`, not inside it

File: kvrocks.conf [hunk 3]
```diff
@@ -643,6 +643,13 @@ migrate-batch-size-kb 16
 # Default: 16M
 migrate-batch-rate-limit-mb 16
 
+
+# Specifies fast shutdown mode
+# If set to yes, we skip blockcache dtor to speed up shutdown
```
Comment (PragmaTwice) [inline kvrocks.conf]:
```suggestion
# If it is set to yes, kvrocks will skip the deallocation of block cache
# while closing the database to speed up the shutdown
```

**Review by PragmaTwice — DISMISSED** [review]

File: tests/gocase/unit/config/config_test.go [hunk 4]
```diff
@@ -320,3 +320,25 @@ func TestGenerateConfigsMatrix(t *testing.T) {
...
+	require.EqualValues(t, "yes", val["skip-block-cache-deallocation-on-close"])
+
+	// default value "no"
+	srv1 := util.StartServer(t, map[string]string{})
+	defer srv1.Close()
+
+	rdb = srv1.NewClient()
+	val = rdb.ConfigGet(ctx, "skip-block-cache-deallocation-on-close").Val()
+	require.EqualValues(t, "no", val["skip-block-cache-deallocation-on-close"])
+}
```
Comment (git-hulk) [inline tests/gocase/unit/config/config_test.go]:
```suggestion
func TestGetConfigSkipBlockCacheDeallocationOnClose(t *testing.T) {
	srv := util.StartServer(t, map[string]string{
		"skip-block-cache-deallocation-on-close": "yes",
	})
	defer srv.Close()

	ctx := context.Background()
	rdb := srv.NewClient()
	defer func() { require.NoError(t, rdb.Close()) }()
	val := rdb.ConfigGet(ctx, "skip-block-cache-deallocation-on-close").Val()
	require.EqualValues(t, "yes", val["skip-block-cache-deallocation-on-close"])

	// default value "no"
	srv1 := util.StartServer(t, map[string]string{})
	defer srv1.Close()

	rdb = srv1.NewClient()
	val = rdb.ConfigGet(ctx, "skip-block-cache-deallocation-on-close").Val()
	require.EqualValues(t, "no", val["skip-block-cache-deallocation-on-close"])
}

```

**Review by git-hulk — DISMISSED** [review]

File: src/storage/storage.cc [hunk 5]
```diff
@@ -88,6 +88,13 @@ Storage::Storage(Config *config)
 Storage::~Storage() {
   DestroyBackup();
   CloseDB();
+  SkipBlockCacheDeallocationOnClose();
```
Comment (mapleFU) [inline src/storage/storage.cc]: nit: you can change to `TrySkip`, since it may not stop this

Comment (wanghenshui) [reply src/storage/storage.cc]: ok

**Review by git-hulk — APPROVED** [review]

**Review by mapleFU — APPROVED** [review]
General LGTM but maybe in the future we can have better method for this, I think this flag can be set only when `cli/main.cc`. But this is also ok to me

**Pull request merged**